File: app/vectorbt_test/engine/portfolio_builder.py
from typing import Dict, List, Any
from vectorbt_test.core.factors import Factor
from vectorbt_test.core.portfolio import PortfolioParameters
from vectorbt_test.portfolios.signal_strategy_portfolio import SignalStrategyPortfolio
from vectorbt_test.portfolios.weight_strategy_portfolio import WeightStrategyPortfolio
from vectorbt_test.portfolios.signal_strategy_portfolio import StrategyOp, SignalStrategy
from vectorbt_test.portfolios.weight_strategy_portfolio import WeightStrategy
from vectorbt_test.core.signals import Signal
from vectorbt_test.core.strategy import StrategyMode
from vectorbt_test.core.portfolio import PortfolioType


class PortfolioBuilder:
    strategies: List[Dict[str, Any]] 
    strategy_weights: List[float] | None
    vote_weights: List[float] | None
    strategy_op: StrategyOp
    schedule_signal: str | Signal
    portfolio_params: PortfolioParameters | None
    _current_strategy: Dict[str, Any] | None
    _current_factor: Factor | str
    
    def __init__(self, name: str, portfolio_mode: str):
        self.name = name
        self.portfolio_mode = portfolio_mode
        self.strategies = []
        self.strategy_weights = None
        self.vote_weights = None
        self.strategy_op = StrategyOp.AND
        self.schedule_signal = None
        self.portfolio_params = None

        self._current_strategy = None
        self._current_factor = None

    @classmethod
    def new(cls, name: str, portfolio_mode: str):
        return cls(name, portfolio_mode)
    
    def add_strategy(self, name: str):
        strategy = {
            "name": name,
            "factors": [],
            "signal": None,
            "strategy_mode": None,
            "top_n": None,
            "threshold": None,
        }
        self.strategies.append(strategy)
        self._current_strategy = strategy
        return self
    
    def add_factor(self, factor: Factor | str):
        if self._current_strategy:
            self._current_strategy["factors"].append(factor)
            self._current_factor = factor
        return self

    def end_factor(self):
        self._current_factor = None
        return self
    
    def set_strategy_signal(self, signal: str | Signal):
        if self._current_strategy:
            self._current_strategy['signal'] = signal
        return self
    
    def set_strategy_mode(self, strategy_mode: str | StrategyMode):
        if self._current_strategy:
            self._current_strategy['strategy_mode'] = StrategyMode(strategy_mode) if isinstance(strategy_mode, str) else strategy_mode
        return self
      
    def end_strategy(self):
        self._current_strategy = None
        return self
    
    def set_schedule_signal(self, signal: str | Signal):
        self.schedule_signal = signal
        return self
    
    def set_strategy_op(self, op: str):
        self.strategy_op = StrategyOp(op.lower() if op is not None else StrategyOp.OR)
        return self
    
    def set_strategy_weights(self, weights):
        self.strategy_weights = weights
        return self
    
    def set_vote_weights(self, weights):
        self.vote_weights = weights
        return self

    def set_portfolio_params(self, params: PortfolioParameters):
        self.portfolio_params = params
        return self
    
    def build(self):
        strategies_obj = []

        for s in self.strategies:
            if PortfolioType(self.portfolio_mode.lower()) == PortfolioType.SIGNAL_STRATEGY:
                strategies_obj.append(
                    SignalStrategy(
                        name=s["name"],
                        factors=s["factors"],
                        signal=s["signal"],
                        threshold=float(s["threshold"]) if s["threshold"] is not None else 0.0
                    )
                )
            else:
                strategies_obj.append(
                    WeightStrategy(
                        name=s["name"],
                        factors=s["factors"],
                        signal=s["signal"],
                        top_n=int(s["top_n"]) if s["top_n"] is not None else 10
                    )
                )

        if PortfolioType(self.portfolio_mode.lower()) == PortfolioType.SIGNAL_STRATEGY:
            return SignalStrategyPortfolio(
                strategies=strategies_obj,
                strategy_op=self.strategy_op,
                schedule_signal=self.schedule_signal,
                vote_weights=self.vote_weights,
                portfolio_params=self.portfolio_params
            )
        else:
            return WeightStrategyPortfolio(
                strategies=strategies_obj,
                strategy_weights=self.strategy_weights,
                schedule_signal=self.schedule_signal,
                portfolio_params=self.portfolio_params
            )

File: quant-mantis/src/components/portfolio/PortfolioBuilder.ts
// src/lib/PortfolioBuilder.ts
// import { Node } from '../../model/dsl_node/nodes'; // 从现有移植代码导入
import { ExprParser } from '../../model/dsl_node/node_expr_parser';

export enum StrategyMode {
  LONG_ONLY = 'long_only',
  SHORT_ONLY = 'short_only',
  LONG_SHORT = 'long_short'
}

export enum StrategyOp {
  AND = 'and',
  OR = 'or'
}

export interface StrategyConfig {
  name: string;
  factors: string[];
  signal?: string;
  strategyMode: StrategyMode;
  threshold?: number;
  topN?: number;
}

export interface PortfolioConfig {
  name: string;
  portfolioMode: 'signal_strategy' | 'weight_strategy';
  strategies: StrategyConfig[];
  strategyOp: StrategyOp;
  scheduleSignal?: string;
  strategyWeights?: number[];
  voteWeights?: number[];
  portfolioParams?: Record<string, any>;
}

export class PortfolioBuilder {
  private config: PortfolioConfig;
  private currentStrategy: StrategyConfig | null = null;

  constructor(name: string, portfolioMode: 'signal_strategy' | 'weight_strategy') {
    this.config = {
      name,
      portfolioMode,
      strategies: [],
      strategyOp: StrategyOp.AND,
    };
  }

  addStrategy(name: string): this {
    this.currentStrategy = {
      name,
      factors: [],
      strategyMode: StrategyMode.LONG_ONLY,
    };
    this.config.strategies.push(this.currentStrategy);
    return this;
  }

  addFactor(factorExpr: string): this {
    if (!this.currentStrategy) throw new Error('No active strategy');
    // 可选：使用ExprParser验证表达式合法性
    try {
      new ExprParser().parse(factorExpr);
    } catch (e) {
      console.warn('Expression validation warning:', e);
    }
    this.currentStrategy.factors.push(factorExpr);
    return this;
  }

  setStrategySignal(signal: string): this {
    if (this.currentStrategy) this.currentStrategy.signal = signal;
    return this;
  }

  setStrategyMode(mode: StrategyMode | string): this {
    if (this.currentStrategy) {
      this.currentStrategy.strategyMode = typeof mode === 'string' 
        ? StrategyMode[mode.toUpperCase() as keyof typeof StrategyMode] 
        : mode;
    }
    return this;
  }

  setStrategyThreshold(threshold: number): this {
    if (this.currentStrategy) this.currentStrategy.threshold = threshold;
    return this;
  }

  setStrategyTopN(topN: number): this {
    if (this.currentStrategy) this.currentStrategy.topN = topN;
    return this;
  }

  endStrategy(): this {
    this.currentStrategy = null;
    return this;
  }

  setScheduleSignal(signal: string): this {
    this.config.scheduleSignal = signal;
    return this;
  }

  setStrategyOp(op: 'and' | 'or'): this {
    this.config.strategyOp = op === 'and' ? StrategyOp.AND : StrategyOp.OR;
    return this;
  }

  setStrategyWeights(weights: number[]): this {
    this.config.strategyWeights = weights;
    return this;
  }

  setVoteWeights(weights: number[]): this {
    this.config.voteWeights = weights;
    return this;
  }

  setPortfolioParams(params: Record<string, any>): this {
    this.config.portfolioParams = params;
    return this;
  }

  build(): PortfolioConfig {
    return this.config;
  }
}